File: crates/flighthq-scene-gl/src/emissive_gl_mesh_material_renderer.rs
//! The built-in Emissive forward renderer.
//!
//! Ports `@flighthq/scene-gl` `emissiveGlMeshMaterialRenderer.ts`. Self-
//! illuminating and lighting-independent: `bind` selects the unlit variant for the
//! emissive map / alpha mode and uploads the linear emissive color scaled by
//! `emissive_strength` (values > 1 drive bloom over the rgba16f scene target).
//! Lights are ignored.

use flighthq_materials::unlit_materials::emissive_material_kind;
use flighthq_render_gl::GlRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::pbr_material::MaterialAlphaMode;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};

use crate::gl_mesh_material_registry::{
    GlMeshMaterialRenderer, MeshMaterial, register_gl_mesh_material_renderer,
};
use crate::gl_mesh_program::{
    begin_gl_mesh_draw, draw_gl_mesh_subset, set_gl_mesh_view_projection,
};
use crate::gl_pbr_standard_bind::{
    bind_gl_pbr_standard_texture, is_gl_texture_ready, unpack_color_to_linear,
};
use crate::gl_scene_runtime::{GlMeshUpload, GlSceneRuntime};
use crate::gl_unlit_prelude::{GlUnlitDefineKey, bind_gl_unlit_surface, ensure_gl_unlit_program};

/// The built-in Emissive renderer.
pub struct EmissiveGlMeshMaterialRenderer;

impl GlMeshMaterialRenderer for EmissiveGlMeshMaterialRenderer {
    fn bind(
        &self,
        state: &mut GlRenderState,
        scene: &mut GlSceneRuntime,
        material: Option<&dyn MeshMaterial>,
        _lights: &SceneLightBlock,
        camera: &Camera,
    ) {
        let emissive = material.and_then(|m| m.as_emissive());
        let key = GlUnlitDefineKey {
            alpha_mask_enabled: emissive.is_some_and(|e| e.alpha_mode == MaterialAlphaMode::Mask),
            has_color_map: emissive.is_some_and(|e| is_gl_texture_ready(e.emissive_map.as_ref())),
            vertex_color: false,
        };
        let program = ensure_gl_unlit_program(state, scene, &key);
        begin_gl_mesh_draw(
            state,
            scene,
            &program.base,
            emissive.is_some_and(|e| e.double_sided),
        );
        set_gl_mesh_view_projection(&state.gl, program.base.loc_view_projection.as_ref(), camera);

        let (color, intensity, alpha_cutoff) = match emissive {
            None => ([1.0, 1.0, 1.0, 1.0], 1.0, 0.5),
            Some(e) => (
                unpack_color_to_linear(e.emissive),
                e.emissive_strength,
                e.alpha_cutoff,
            ),
        };
        bind_gl_unlit_surface(state, &program, &color, intensity, alpha_cutoff);
        bind_gl_pbr_standard_texture(
            state,
            emissive.and_then(|e| e.emissive_map.as_ref()),
            program.loc_color_map.as_ref(),
            0,
        );
    }

    fn draw(
        &self,
        state: &mut GlRenderState,
        scene: &mut GlSceneRuntime,
        proxy: &SceneRenderProxy,
        upload: &GlMeshUpload,
    ) {
        draw_gl_mesh_subset(state, scene, proxy, upload);
    }
}

/// Registers the built-in Emissive renderer for `EmissiveMaterialKind`.
pub fn register_emissive_gl_material(scene: &mut GlSceneRuntime) {
    register_gl_mesh_material_renderer(
        scene,
        emissive_material_kind(),
        Box::new(EmissiveGlMeshMaterialRenderer),
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use flighthq_materials::unlit_materials::emissive_material_kind;

    use crate::gl_mesh_material_registry::get_gl_mesh_material_renderer;
    use crate::gl_scene_runtime::create_gl_scene_runtime;

    // register_emissive_gl_material

    #[test]
    fn register_emissive_gl_material_registers_for_the_emissive_kind() {
        let mut scene = create_gl_scene_runtime();
        register_emissive_gl_material(&mut scene);
        assert!(get_gl_mesh_material_renderer(&scene, emissive_material_kind()).is_some());
    }
}

File: packages/scene-gl/src/emissiveGlMeshMaterialRenderer.ts
import { unpackColorToLinear } from '@flighthq/color';
import type { LinearColor } from '@flighthq/types';
import type {
  Camera3D,
  EmissiveMaterial,
  GlMeshMaterialRenderer,
  GlRenderState,
  Material,
  MeshGeometry,
  SceneLightBlock,
  SceneRenderProxy,
} from '@flighthq/types';
import { EmissiveMaterialKind } from '@flighthq/types';

import { registerGlMeshMaterialRenderer } from './glMeshMaterialRegistry';
import {
  beginGlMeshDraw,
  bindGlUvTransform,
  drawGlMeshSubset,
  hasGlUvTransform,
  setGlMeshViewProjection,
} from './glMeshProgram';
import { getGlSceneRuntime } from './glSceneRuntime';
import type { GlUnlitDefineKey } from './glUnlitPrelude';
import { bindGlUnlitSurface, ensureGlUnlitProgram } from './glUnlitPrelude';

// The built-in Emissive forward renderer (GlMeshMaterialRenderer for EmissiveMaterialKind). Self-
// illuminating and lighting-independent: bind selects the unlit variant for the emissive map / alpha
// mode and uploads the linear emissive color scaled by emissiveStrength (values > 1 drive bloom over
// the rgba16f scene target). Lights are ignored. See registerEmissiveGlMaterial to install it.
export const emissiveGlMeshMaterialRenderer: GlMeshMaterialRenderer = {
  bind(
    state: GlRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera3D>,
  ): void {
    const gl = state.gl;
    const emissive = material as Readonly<EmissiveMaterial> | null;
    const program = ensureGlUnlitProgram(state, defineKeyForMaterial(emissive));
    beginGlMeshDraw(state, program, emissive !== null && emissive.doubleSided);
    setGlMeshViewProjection(gl, program.locViewProjection, camera);

    if (emissive === null) {
      bindGlUnlitSurface(state, program, WHITE, 1, null, 0.5);
      return;
    }
    unpackColorToLinear(scratchRgba, emissive.emissive);
    bindGlUnlitSurface(
      state,
      program,
      scratchRgba,
      emissive.emissiveStrength,
      emissive.emissiveMap,
      emissive.alphaCutoff,
    );
    bindGlUvTransform(gl, program, emissive.emissiveMap);
  },

  draw(state: GlRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    const program = getGlSceneRuntime(state).activeMeshProgram;
    if (program === null) return;
    drawGlMeshSubset(state, program, proxy, geometry);
  },
};

// Registers the built-in Emissive renderer for EmissiveMaterialKind on this state. Opt-in (no top-
// level side effect); call once per GlRenderState before drawScene so meshes with EmissiveMaterials
// draw.
export function registerEmissiveGlMaterial(state: GlRenderState): void {
  registerGlMeshMaterialRenderer(state, EmissiveMaterialKind, emissiveGlMeshMaterialRenderer);
}

function defineKeyForMaterial(material: Readonly<EmissiveMaterial> | null): GlUnlitDefineKey {
  return {
    alphaMaskEnabled: material !== null && material.alphaMode === 'mask',
    hasColorMap: material !== null && material.emissiveMap !== null && material.emissiveMap.image !== null,
    hasUvTransform: hasGlUvTransform(material !== null ? material.emissiveMap : null),
    vertexColor: false,
  };
}

const scratchRgba: LinearColor = [0, 0, 0, 0];
const WHITE: LinearColor = [1, 1, 1, 1];
